sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs

```diff
@@ -57,42 +58,93 @@ protected override void Initialize(SonarAnalysisContext context)
-                    }
-
-                    var hasIrregularPattern = decimalParts
-                        .SelectMany(part => part.Split('_'))
-                        .Select(x => x.Length)
-                        .Skip(1) // skip the first part (1_234 => 234)
-                        .Reverse().Skip(decimalParts.Length == 2 ? 1 : 0) // skip the last if there is a decimal (.234_5 => 234)
-                        .Distinct().Skip(1).Any(); // we expect to have only 1 size of pattern
-
-                    if (hasIrregularPattern)
+                    if (HasIrregularPattern(literal.Token.Text))
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, literal.GetLocation()));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        public static bool HasIrregularPattern(string numericToken)
         {
-            if (numberLiteral.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            // hexadecimal en binary prefixes
+            var start = numericToken.StartsWith("0x", StringComparison.InvariantCultureIgnoreCase)
+                || numericToken.StartsWith("0b", StringComparison.InvariantCultureIgnoreCase) ? 2 : 0;
+
+            var length = numericToken.Length - start;
+
+            // UL suffix.
+            if (numericToken.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            {
+                length -= 2;
+            }
+            // single suffixes
+            else if ("LDFUMldfum".IndexOf(numericToken[numericToken.Length - 1]) != NotFound)
+            {
+                length--;
+            }
+
+            var stripped = numericToken.Substring(start, length);
+            var splitted = stripped.Split(Dot);
+
+            // ignore multiple dots.
+            if (splitted.Length > 2)
+            {
+                return false;
+            }
+
+            var groups = splitted[0].Split(Underscore).Select(s => s.Length).ToArray();
+
+            var size = NotFound;
+
+            if (groups.Length > 1)
+            {
+                size = groups[1];
+
+                // first should not be bigger.
+                if (groups[0] > size)
+                {
+                    return true;
+                }
+
+                // all should have the same size, except the first.
+                if (groups.Skip(1).Any(g => g != size))
+                {
+                    return true;
+                }
+            }
+
+            if (splitted.Length == 1)
+            {
+                return false;
+            }
+
+            var decimals = splitted[1].Split(Underscore).Select(s => s.Length).ToArray();
+
+            if (decimals.Length == 1)
+            {
+                return false;
+            }
+
+            // The first decimal group should be one smaller.
+            decimals[0]++;
```

Comment:
Why this change in the rule behaviour? I'm not saying this is necessarily wrong, but if we change that, it first should be updated in the rule description

Reply:
I think is the only right implementation. Given this changes, some descriptions have to be updated anyway.

Reply:
Could you direct me to some documentation regarding this ? I cannot find anything on the grouping of the fractional part to justify this change

Reply:
@christophe-zurn-sonarsource No, I can't. :(

I'm pretty sure I've been educated that way, but I can't find any source that says I'm right.

it is probably only true if the group size is 3 anyway. What I remember is that the first decimal are 1/10, then 1/100, so it makes sense to add a _ between 1/100 and 1/1000, just like between 1000 and 100.

But I agree, that without proof, we should not go for it.

The only thing I found (that also slightly disagrees) is this:
https://math.stackexchange.com/questions/182775/convention-of-digit-grouping-after-decimal-point/182776#182776

What suggests that the group size of your decimal part can be different (although it still should be consequent).

Reply:
I've changed the behaviour. Also I added that the group size should be at least, two, as I think that makes sense. ;)